core/tree/multi_root_tree.go

```diff
@@ -61,36 +65,74 @@ func (t *MultiRootTree) IsCyclic() error {
 	return nil
 }
 
+// get sorted dataMap keys
+func (t *MultiRootTree) getSortedKeys() []string {
+	keys := []string{}
+	for k := range t.dataMap {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+func (t *MultiRootTree) isCyclic(root *TreeNode, visited map[string]bool) error {
```

Comment:
IMHO, we can implement function within function to achieve recursion as required. With this approach, the caller of the main function, perhaps in this case hasCycle, does not need to think about how to initiate pathMap or any other arguments. Ideally, hasCycle should only accept one argument, which is root, and return only one result, which is error. Though, again, we may need to reconsider whether to use hasCycle as the function name. Because, when I read hasCycle, I would assume the return value is boolean. Similar to isValid, which is boolean and Validate which returns error. Maybe we can use validateCyclic or any other name that express more intent.

Reply:
IMHO, we can implement function within function to achieve recursion as required. With this approach, the caller of the main function, perhaps in this case `hasCycle`, does not need to think about how to initiate `pathMap` or any other arguments. Ideally, `hasCycle` should only accept one argument, which is `root`, and return only one result, which is `error`. Though, again, we may need to reconsider whether to use `hasCycle` as the function name. Because, when I read `hasCycle`, I would assume the return value is boolean. Similar to `isValid`, which is boolean and `Validate` which returns error. Maybe we can use `validateCyclic` or any other name that express more intent.